biz.aQute.bndlib/src/aQute/bnd/service/result/Result.java

```diff
@@ -0,0 +1,162 @@
+ */
+public interface Result<V, E> {
+
+	/**
+	 * Returns the value of this instance as an {@link Optional}. Returns
+	 * Optional.empty() if this is an Err instance.
+	 *
+	 * @return see above.
+	 */
+	Optional<V> getValue();
+
+	/**
+	 * Returns the error of this instance as an {@link Optional}. Returns
+	 * Optional.empty() if this is an Ok instance.
+	 *
+	 * @return see above.
+	 */
+	Optional<E> getError();
+
+	/**
+	 * Returns <code>true</code> if this instance represents an Ok value, false
+	 * otherwise.
+	 *
+	 * @return see above.
+	 */
+	boolean isOk();
+
+	/**
+	 * Returns <code>true</code> if this instance represents an Err value, false
+	 * otherwise.
+	 *
+	 * @return see above
+	 */
+	boolean isErr();
+
+	/**
+	 * Returns the contained value if this is an Ok value, otherwise throws a
+	 * ResultException.
+	 *
+	 * @return the contained value
+	 * @throws ResultException in case unwrap() is called on an Err value
+	 */
+	V unwrap() throws ResultException;
+
+	/**
+	 * Express the expectation that this object is an Ok value. If it's an Err
+	 * value instead, throw a ResultException with the given message.
+	 *
+	 * @param message the message to pass to a potential ResultException
+	 * @throws ResultException if unwrap() is called on an Err value
+	 */
+	V unwrap(String message) throws ResultException;
+
+	/**
+	 * If this is an Ok value, andThen() returns the result of the given
+	 * {@link Function}. Otherwise returns this.
+	 *
+	 * @param lambda The {@link Function} to be called with the value of this.
+	 * @param <U> The new value type.
+	 * @return see above.
+	 */
+	@SuppressWarnings("unchecked")
+	<U> Result<U, E> then(final FunctionWithException<V, Result<U, E>> lambda);
+
+	/**
+	 * If this is an Ok value, map() returns the result of the given
+	 * {@link Function}, wrapped in a new Ok Result instance. Otherwise returns
+	 * this.
+	 *
+	 * @param lambda The {@link Function} to call with the value of this.
+	 * @param <U> The new value type.
+	 * @return see above.
+	 */
+	<U> Result<U, E> map(final FunctionWithException<V, U> lambda);
+
+	/**
+	 * If this is an Err value, mapErr() returns the result of the given @{link
+	 * Function}, wrapped in a new Err Result instance. Otherwise returns this.
+	 *
+	 * @param lambda The {@link Function} to call with the error of this.
+	 * @param <F> The new error type.
+	 * @return see above
+	 */
+	<F> Result<V, F> mapErr(final FunctionWithException<E, F> lambda);
+
+	/**
+	 * If this is an ok value, return this. Otherwise change the error into a
+	 * new value and return ok.
+	 *
+	 * @param recover lambda converting error in a value
+	 * @return a result
+	 */
+	Result<V, E> recover(FunctionWithException<E, V> recover);
+
+	default V orElse(V value) {
+		return getValue().orElse(value);
+	}
+
+	default V orElseGet(Supplier<V> value) {
+		return getValue().orElseGet(value);
+	}
+
+	<R extends Exception> V orElseThrow(FunctionWithException<E, R> f) throws Exception;
```

Comment:
This should be `<R extends Throwable>` to allow the function to make an `Error`.

Reply:
Sure? That means the method must throws Throwable?

Reply:
```java
<R extends Throwable> V orElseThrow(FunctionWithException<? super E, ? extends R> exceptionSupplier) throws R 
```

The API also needs some overall `? super`/`? extends` love to make it more useful.

Reply:
I was secretly hoping you could put this magic on? I've added this one but you got way more knowledge in this area than I will ever master.

Reply:
![image](https://user-images.githubusercontent.com/200494/69440110-7c27a200-0d48-11ea-98b9-603e3b2cf982.png)

Reply:
LOL. I can do this once this PR makes it in.

Reply:
![image](https://user-images.githubusercontent.com/200494/69440151-8fd30880-0d48-11ea-85af-8ad5e8ef7c70.png)

Reply:
The problem is the use of FunctionWithException which itself can throw an Exception which may different than R :-) So you may need to cast the result of apply to `(R)` to make the compiler shut up. Or catch any exception from the apply method and throw it in a ResultException.

Reply:
Something like:

```java
R r;
try {
  r = f.apply(error);
} catch (Exception e) {
  throw new ResultException(e);
}
throw r;

Reply:
There is also the risk the function can return null which should be handled somehow to make the API robust.

Reply:
```java
R r;
try {
  r = f.apply(error);
} catch (Exception e) {
  throw new ResultException(e);
}
if (r != null) {
  throw r;
}
throw new ResultException(error);